internal/access/destinationcredential.go

```diff
@@ -0,0 +1,195 @@
+package access
+
+import (
+	"database/sql"
+	"fmt"
+	"time"
+
+	"github.com/gin-gonic/gin"
+
+	"github.com/infrahq/infra/api"
+	"github.com/infrahq/infra/internal"
+	"github.com/infrahq/infra/internal/logging"
+	"github.com/infrahq/infra/internal/server/data"
+	"github.com/infrahq/infra/internal/server/models"
+	"github.com/infrahq/infra/uid"
+)
+
+type ListDestinationCredentialResponse struct {
+	Items          []models.DestinationCredential
+	MaxUpdateIndex int64
+}
+
+func (l ListDestinationCredentialResponse) UpdateIndex() int64 {
+	return l.MaxUpdateIndex
+}
+
+func (l ListDestinationCredentialResponse) ItemCount() int {
+	return len(l.Items)
+}
+
+func CreateDestinationCredential(c *gin.Context, destination string) (*models.DestinationCredential, error) {
+	rCtx := GetRequestContext(c)
+	// does the user have a grant to this destination?
+	// ListGrants(c, )
+	grants, err := data.ListGrants(rCtx.DBTxn, data.ListGrantsOptions{
+		ByDestination:              destination,
+		BySubject:                  uid.NewIdentityPolymorphicID(rCtx.Authenticated.User.ID),
+		IncludeInheritedFromGroups: true,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("ListGrants: %w", err)
+	}
+	if len(grants) == 0 {
+		return nil, internal.ErrUnauthorized
+	}
+
+	dest, err := data.GetDestination(rCtx.DBTxn, data.GetDestinationOptions{ByName: destination})
+	if err != nil {
+		return nil, fmt.Errorf("GetDestination: %w", err)
+	}
+
+	cr := &models.DestinationCredential{
+		ID:                 uid.New(),
+		OrganizationMember: models.OrganizationMember{OrganizationID: rCtx.Authenticated.User.OrganizationID},
+		RequestExpiresAt:   time.Now().Add(2 * time.Minute),
+		UserID:             rCtx.Authenticated.User.ID,
+		DestinationID:      dest.ID,
+	}
+	err = data.CreateDestinationCredential(rCtx.DBTxn, cr)
+	if err != nil {
+		return nil, fmt.Errorf("CreateDestinationCredential: %w", err)
+	}
+	logging.Debugf("Creating DestinationCredential with id %d, orgID %d", cr.ID, cr.OrganizationID)
+
+	// commit the transaction so others can see the request
+	err = rCtx.DBTxn.Commit()
+	if err != nil {
+		return nil, err
+	}
+
+	// wait for the request to be filled; listen for the notify on the specific record req.ID
+	listener, err := data.ListenForNotify(c, rCtx.DataDB, data.ListenForNotifyOptions{
+		OrgID:                      rCtx.Authenticated.AccessKey.OrganizationID,
```

Comment:
I feel like Org ID should come from the transaction context rather than something that can be specified